File: RainMusic/ViewModels/PlayerViewModel.swift
import SwiftUI

@Observable
class PlayerViewModel {
    private let audioManager = AudioPlayerManager.shared
    
    var currentSong: Song? { audioManager.currentSong }
    var isPlaying: Bool { audioManager.isPlaying }
    var currentTime: Double { audioManager.currentTime }
    var duration: Double { audioManager.duration }
    var volume: Float { audioManager.volume }
    var shuffleMode: ShuffleMode { audioManager.shuffleMode }
    var repeatMode: RepeatMode { audioManager.repeatMode }
    var lyrics: [LyricLine] { audioManager.lyrics }
    var currentLyricIndex: Int { audioManager.currentLyricIndex }
    var showLyrics: Bool { audioManager.showLyrics }
    var playlist: [Song] { audioManager.playlist }
    var currentIndex: Int { audioManager.currentIndex }
    
    var currentTimeFormatted: String { audioManager.currentTimeFormatted }
    var durationFormatted: String { audioManager.durationFormatted }
    
    var progress: Double {
        guard duration > 0 else { return 0 }
        return currentTime / duration
    }
    
    init() {}
    
    func play(song: Song) {
        audioManager.play(song: song)
    }
    
    func togglePlayPause() {
        audioManager.togglePlayPause()
    }
    
    func playNext() {
        audioManager.playNext()
    }
    
    func playPrevious() {
        audioManager.playPrevious()
    }
    
    func seek(to time: Double) {
        audioManager.seek(to: time)
    }
    
    func setVolume(_ volume: Float) {
        audioManager.setVolume(volume)
    }
    
    func setPlaylist(_ songs: [Song], startIndex: Int = 0) {
        audioManager.setPlaylist(songs, startIndex: startIndex)
    }
    
    func toggleShuffle() {
        audioManager.toggleShuffle()
    }
    
    func cycleRepeatMode() {
        audioManager.cycleRepeatMode()
    }
    
    func toggleLyrics() {
        audioManager.toggleLyrics()
    }
    
    func updateLyrics() {
        audioManager.updateLyrics()
    }
    
    func formatTime(_ time: Double) -> String {
        audioManager.formatTime(time)
    }
}

File: RainMusicAndroid/app/src/main/java/com/rain/music/viewmodel/PlayerViewModel.kt
package com.rain.music.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.rain.music.data.db.MusicDatabase
import com.rain.music.data.model.Song
import com.rain.music.manager.*
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

class PlayerViewModel(application: Application) : AndroidViewModel(application) {

    private val audioPlayerManager = AudioPlayerManager(application)
    private val songDao = MusicDatabase.getInstance(application).songDao()

    // 播放状态
    val currentSong = audioPlayerManager.currentSong
    val isPlaying = audioPlayerManager.isPlaying
    val currentTime = audioPlayerManager.currentTime
    val duration = audioPlayerManager.duration
    val shuffleMode = audioPlayerManager.shuffleMode
    val repeatMode = audioPlayerManager.repeatMode
    val systemVolume = audioPlayerManager.systemVolume

    // 歌词
    private val _lyrics = MutableStateFlow<List<LyricLine>>(emptyList())
    val lyrics: StateFlow<List<LyricLine>> = _lyrics

    private val _currentLyricIndex = MutableStateFlow(-1)
    val currentLyricIndex: StateFlow<Int> = _currentLyricIndex

    private val _showLyrics = MutableStateFlow(false)
    val showLyrics: StateFlow<Boolean> = _showLyrics

    init {
        audioPlayerManager.initialize()

        // 监听歌曲变化加载歌词
        viewModelScope.launch {
            currentSong.collect { song ->
                if (song != null) {
                    _lyrics.value = LyricsParser.loadLyrics(song) ?: emptyList()
                    _currentLyricIndex.value = -1
                }
            }
        }

        // 更新歌词索引
        viewModelScope.launch {
            currentTime.collect { time ->
                if (_lyrics.value.isNotEmpty()) {
                    _currentLyricIndex.value = LyricsParser.findCurrentLineIndex(_lyrics.value, time)
                }
            }
        }
    }

    // 播放控制
    fun play(song: Song) = audioPlayerManager.play(song)
    fun togglePlayPause() = audioPlayerManager.togglePlayPause()
    fun seekTo(position: Long) = audioPlayerManager.seekTo(position)
    fun setVolume(volume: Float) = audioPlayerManager.setVolume(volume)

    fun setPlaylist(songs: List<Song>, startIndex: Int = 0) {
        audioPlayerManager.setPlaylist(songs, startIndex)
    }

    fun playNext(songs: List<Song> = emptyList()) {
        if (audioPlayerManager.playlist.isEmpty() && songs.isNotEmpty()) {
            val currentIndex = songs.indexOfFirst { it.id == currentSong.value?.id }
            audioPlayerManager.setPlaylist(songs, if (currentIndex >= 0) currentIndex else 0)
            return
        }
        audioPlayerManager.playNext()
    }

    fun playPrevious(songs: List<Song> = emptyList()) {
        if (audioPlayerManager.playlist.isEmpty() && songs.isNotEmpty()) {
            val currentIndex = songs.indexOfFirst { it.id == currentSong.value?.id }
            audioPlayerManager.setPlaylist(songs, if (currentIndex >= 0) currentIndex else 0)
            return
        }
        audioPlayerManager.playPrevious()
    }

    fun toggleShuffle() = audioPlayerManager.toggleShuffle()
    fun cycleRepeatMode() = audioPlayerManager.cycleRepeatMode()
    fun toggleLyrics() { _showLyrics.value = !_showLyrics.value }

    // 更新封面
    fun updateAlbumArt(songId: Long, newArtUri: String) {
        viewModelScope.launch {
            val song = songDao.getSongById(songId)
            if (song != null) {
                song.albumArtUri?.let { oldUri ->
                    try { java.io.File(oldUri).delete() } catch (_: Exception) {}
                }
                songDao.update(song.copy(albumArtUri = newArtUri))
            }
        }
    }

    // 格式化时间
    fun formatTime(ms: Long): String {
        val totalSeconds = ms / 1000
        val minutes = totalSeconds / 60
        val seconds = totalSeconds % 60
        return String.format("%d:%02d", minutes, seconds)
    }

    override fun onCleared() {
        super.onCleared()
        audioPlayerManager.release()
    }
}
